[random_fp] Random numbers generation with properties

jposwiata opened the pull request:
new_random_xxx() family extended to configure created numbers. Generation is handled with overloaded macro(s) configured with flags defined in RANDOM_GEN_FLAGS enumeration.
Default implementation keep slow version of preparing chains of bits (no functional chages to existing tests, the values are exactly identical for "old" seed values).
All classes follows the same pattern, unified predicates are available for both C and C++, most functions are defined as inlined (except from/to non-trivial conversions).
Code prepared with TDD, "C" tests do not verify all functions in details (no support from GTest for "C" language).

The PR depends on #945 and #938 (with `random_bits` implementations), commits included here.

AthenasJimenez reviewed (changes requested):
Initial thoughts, I still have pending to review `Floats.cpp`.

jposwiata commented on framework/unit-tests/sandstone_utils_tests.cpp:
```diff
@@ -539,3 +561,210 @@ TEST(FloatConversions, HF8toFloat) {
+    // .to_float() gives more precise result, TO_FLOAT always reports float!
+    EXPECT_EQ(TO_FLOAT(VECTOR_FLOAT64), (float) new_random_float64(STATIC_VECTOR).to_float());
+    EXPECT_EQ(TO_FLOAT(VECTOR_FLOAT64), (float) new_random_double(STATIC_VECTOR));
+    EXPECT_EQ(TO_FLOAT(VECTOR_FLOAT80), (float) new_random_float80(STATIC_VECTOR).to_float());
```
Precission loss is expected as pointed out in the comment above. The code expects that "compiler" converted values are exactly the same.

jposwiata commented on framework/fp_vectors/Floats.h:
```diff
@@ -259,24 +280,31 @@ struct HFloat8
     constexpr inline bool is_denormal() const { return (exponent == HFLOAT8_DENORM_EXPONENT) && (mantissa != 0); }
     constexpr inline bool is_valid() const    { return (value != HFLOAT8_NAN_INF_VALUE) && (value != HFLOAT8_SATURATED_OVERFLOW_VALUE); }
-    constexpr inline bool is_nan_inf() const  { return value == HFLOAT8_NAN_INF_VALUE; }
+    constexpr inline bool is_inf_nan() const  { return value == HFLOAT8_NAN_INF_VALUE; }
```
As the type definitions are not consistent, we cannot keep the consistency here.
HFloat8 is a bit different semantics: it doesn't include separate Inf and NaN values. In fact there's no detailed info if the data was out of range (e.g. big * big) or if it was wrong (e.g. anything / 0).
I think I should take another approach for this particular type (for Inf/NaN both IS_INF() and IS_NAN() should report true). The question is what should report IS_SNAN/IS_QNAN..

AthenasJimenez replied:
My comment was more related to `inf_nan` vs `nan_inf`. My request is to chose one and keep it for all implementations when applicable, including macros and functions/methods.

jposwiata commented on framework/fp_vectors/Floats.h:
```diff
@@ -464,31 +544,37 @@ struct BFloat16
+    constexpr inline bool is_negative() const { return sign != 0; }
+    constexpr inline bool is_zero() const     { return (exponent == BFLOAT16_DENORM_EXPONENT) && (mantissa == 0); }
+    constexpr inline bool is_denormal() const { return (exponent == BFLOAT16_DENORM_EXPONENT) && (mantissa != 0); }
+    constexpr inline bool is_valid() const    { return exponent != BFLOAT16_NAN_EXPONENT; }
```
This function is intended to check if it is valid value, just to make Inf/NaN/sNaN/qNaN confusion gone.

jposwiata commented on framework/fp_vectors/Floats.h:
```diff
@@ -592,27 +779,102 @@ struct Float80 {
 
+STATIC_INLINE long double from_float80(Float80 f) { return f.as_float; }
+
+#ifdef __SIZEOF_FLOAT128__
```
The type is not used. It has been moved from antoher file just to keep things consistent. I don't know the reason the type was added.
I don't think we will test "these" types and they are not supported by the architecture.

AthenasJimenez replied:
I've never seen/use neither Float80 nor Float128, so I'm ok leaving them as they are.

AthenasJimenez reviewed (changes requested):
Another partial review. Sorry for making it like this but it is a big change.

jposwiata commented on framework/Floats.cpp:
```diff
@@ -304,56 +303,463 @@ float from_hfloat8_emulated(HFloat8 f) {
+}
+
+template<typename F>
+F normalize_impl(F f, float v1, float v2) {
```
As discussed with Łukasz.. The result from the normalization should always and unconditionally be in the given range. If you need any other behaviour, you'll need just to implement your own "normalization" (in the context of test content).
The rule should apply to "special" condition as well: Inf/NaNs should not be normalized. Forced normalization of values not following general rule is considered as development issue and should not be checked in the code (e.g. requesting RANDOM_GEN_FLAGS_FLOAT_SIGN_NEGATIVE should not be specified to avoid confusion).
I introduced additional feature to cover "often used" 0/+-std::numeric__limits<>::max() ranges.

jposwiata replied:
> Agree. My suggestions was intended to allow negative values on range (i.e v1=-5, v2=5). I honesty don't remember previous implementation, but as far as I can see it is implemented so all good.

It is fully allowed and handled properly. Such pattern is not the best in the scope of values precission, you must be aware of that. What more, this doesn't cover denormals/etc handling (which might be "a kind of" disadventage in the scope of particular type/calculation).

> Not an expert, but what would happen for example with Float64? Probably consider a Double? I don't know :/

More distinct values will be available, it makes more sense to run the test on bigger number of input values. Otherwise the calculations are run on exatly the same data.

AthenasJimenez reviewed (dismissed):
Another partial review, I still have pending `gen_random_impl()`.

jposwiata commented on framework/fp_vectors/Floats.h:
```diff
@@ -677,91 +967,439 @@ STATIC_INLINE Float80 new_float80(uint64_t sign, uint64_t exponent, uint64_t jbi
+    RANDOM_GEN_FLAGS_EXTRA_FAST_MEMSET        = 0x10000000, // value initialized with "memset_random()"
+    RANDOM_GEN_FLAGS_FLOAT_FORCE_VALID        = 0x20000000, // valid number, Inf/NaN/overflow excluded
+
+    // short aliases
```
Exactly. The only "exceptions" are full sets (I mean PATTERNED and STATIC_VECTOR) to address the bahaviour of "older" implementations (with their limitations and identical RND/seed handling).

AthenasJimenez commented on framework/Floats.cpp:
```diff
@@ -304,56 +303,538 @@ float from_hfloat8_emulated(HFloat8 f) {
+    // normalize input value to the range, just by normalizing it to 1..2 (with flat distribution)
+    // and then forcing requested range. It will report very few values (esp. for 8bit types),
+    // so it would be needed to normalize e.g. float and downconvert it to appropriate type if
+    // more values is expected, e.g: BFloat8(gen_random_float(10, 20))
```
I tried to use the exact same code in an experiment for a test and got this error:

```
../tests/ardilla/random_floats.cpp: In function 'int random_floats_init(test*)':
../tests/ardilla/random_floats.cpp:64:44: error: too many arguments to function 'float gen_random_float(uint32_t)'
   64 |     BFloat16 bf = BFloat16(gen_random_float(10, 20));
      |                            ~~~~~~~~~~~~~~~~^~~~~~~~
```

jposwiata replied:
Hmmm. I'm a bit confused, it works for me.
Added both (I mean '{}' and '()' constructions) to unit test

AthenasJimenez replied:
Let's discuss offline.

jposwiata commented on framework/sandstone_data.h:
```diff
@@ -229,8 +235,11 @@ template<> struct TypeToDataType<fp16_t> : TypeToDataType_helper<Float16Data> {}
 {
     constexpr unsigned SizeMask = 0x3f;
+    // exceptions
+    if ((type == HFloat8Data) || (type == BFloat8Data))
+        return 1;
     if (type == BFloat16Data)
-        return 2;               // exception
+        return 2;
```
Let have another PR to fix this "hotfix" to follow your comment (there's PR #772 for it).
Can you agree with such way?

jposwiata commented on framework/fp_vectors/Floats.h:
```diff
@@ -677,91 +967,443 @@ STATIC_INLINE Float80 new_float80(uint64_t sign, uint64_t exponent, uint64_t jbi
+
+// template to handle bits in sign/exponent/mantissa according given T
+template<typename T>
+inline T gen_random_tmpl(uint32_t flags) {
```
no one can overload with return type..

jposwiata commented on framework/Floats.cpp:
```diff
@@ -304,56 +303,558 @@ float from_hfloat8_emulated(HFloat8 f) {
+
+static constexpr int BITS_FROM_RANDOM = 31;
+
+thread_local uint32_t random_bits_value = 0;
+thread_local int random_bits_available = 0;
```
if it was used in MT (ti in `run()`), the result are unpredictable (including #UB on shifts).

jposwiata commented on framework/Floats.cpp:
```diff
@@ -304,56 +303,558 @@ float from_hfloat8_emulated(HFloat8 f) {
+    if (bits <= 0) {
+        return I{ 0 };
+    }
+    if (bits >= sizeof(I) * 8) {
+        return (I) (~0);
+    }
+    return (static_cast<I>(1) << bits) - 1;
```
This is not on the code being generated. The compiler/coverity loves to complain a lot on these functions (and.. the list of warnings differs with compiler/version..). I haven't found any "ideal" pattern yet.
Here (as unsigned int types are only used) all the casts might be dropped (and the question is whether it will be used in "signed" context at all..). I even consider to make these a template<typename I, size_t bits> instead (with static assertions), as gave up any plans to have it shared.

jposwiata commented on framework/Floats.cpp:
```diff
@@ -304,56 +303,558 @@ float from_hfloat8_emulated(HFloat8 f) {
+        random_val_value = random32();
+        random_val_available = (1ULL << BITS_FROM_RANDOM);
+    }
+    uint32_t val = random_val_value % range;
+    random_val_available /= range;
+    random_val_value /= range;
+    return val;
```
The point here was not to get flat distribution. Just to reduce RNG usage a bit.
In general flat distribution might be very difficult with our "random" RNGs (either LCG or AES). Especially LCG is very weak with very first values taken (and what worse for some `num`).
I would like to have shared solution in Random.h one day, but "bullet-proof" solution doesn't exist yet. Even "a little better" implementation is out of the scope of this commit.

jposwiata commented on framework/Floats.cpp:
```diff
@@ -304,56 +303,558 @@ float from_hfloat8_emulated(HFloat8 f) {
+    f.exponent = get_mask<typename F::base_type>(F::exponent_bits());
+    if (quiet) {
+        f.mantissa |= static_cast<F::base_type>(1ULL << (F::mantissa_bits() - 1));
+    } else if (f.mantissa == 0) {
+        // prevent against 0 by setting any bit, excluding most significant one (quiet bit)
+        uint32_t bit = get_random_value<uint32_t>(F::mantissa_bits() - 1);
+        f.mantissa = static_cast<F::base_type>(1ULL << bit);
+    } else {
+        // make sure quiet bit is cleared
+        f.mantissa &= static_cast<F::base_type>(~(1ULL << (F::mantissa_bits() - 1)));
+    }
```
For most types there's no cannonical form of these. The mantissa must be non-zero. If 0 randomly selected or forced, at least one bit must be set (let you think about use cases with Constatnt:0 generator..)

jposwiata commented on framework/Floats.cpp:
```diff
@@ -304,56 +303,558 @@ float from_hfloat8_emulated(HFloat8 f) {
-    f.jbit = 1;
-    f.mantissa = set_random_bits(random32() % (FLOAT80_MANTISSA_BITS + 1), FLOAT80_MANTISSA_BITS);
+template<typename F>
+F fix_bits(F f) {
     return f;
 }
+template<>
+Float80 fix_bits(Float80 f) {
+    f.jbit = (f.exponent == 0) ? 0 : 1;
+    return f;
+}
```
I was asked to use template specializations (without constexpr `if`s). Let this code follow this pattern in all places.

jposwiata commented on framework/fp_vectors/Floats.h:
```diff
@@ -592,27 +795,114 @@ struct Float80 {
+    Float128() = default;
+    Float128(long double f) : payload(f) {}
+
+    static constexpr int digits = 113;
+    static constexpr int digits10 = 33;
+    static constexpr int max_digits10 = 6;  // log2(digits)
+    static constexpr int min_exponent = -16381;
+    static constexpr int min_exponent10 = -4931;
+    static constexpr int max_exponent = 16384;
+    static constexpr int max_exponent10 = 4932;
```
For sure this change is out of scope of this PR (I haven't even analyzed "what is it required for").

jposwiata commented on framework/fp_vectors/Floats.h:
```diff
@@ -8,11 +8,22 @@
+#include <immintrin.h>
+#endif
+
+#ifdef __cplusplus
+#define STATIC_INLINE static inline constexpr
+#else
+#define STATIC_INLINE static inline
+#endif
```
These were existing definitions. I wondered why these are here, but haven't pay too much attention to it (I rather replace it with separate "flags" to see what is going on..)

jposwiata commented on framework/fp_vectors/Floats.h:
```diff
@@ -678,91 +968,443 @@ STATIC_INLINE Float80 new_float80(uint64_t sign, uint64_t exponent, uint64_t jbi
-} // extern "C"
+extern "C" {
 #endif
+HFloat8  gen_random_hfloat8 (uint32_t flags);
+BFloat8  gen_random_bfloat8 (uint32_t flags);
+Float16  gen_random_float16 (uint32_t flags);
+BFloat16 gen_random_bfloat16(uint32_t flags);
+Float32  gen_random_float32 (uint32_t flags);
+float    gen_random_float   (uint32_t flags);
+Float64  gen_random_float64 (uint32_t flags);
+double   gen_random_double  (uint32_t flags);
+Float80  gen_random_float80 (uint32_t flags);
```
array/vector/... generation is planned as a separate PR. I didn't plan to change the "main" interface (this is `new_random_float()` function should stay, only the prototype has changed to overloaded "form").
Do you think that using strategies instead of bitfieds selectors (I think 5 lambdas will do) will be significantly better (the code will have yet another tons of templates).

jposwiata replied:
I run the experiments with two implementations: it seems that the compiler is able to optimize the code far better if "inlined" than if appropriate strategies are involved. Single value  a time with fixups is ~20% slower than just single with swich/cases, big array "degrades" to more-or-less same time (95%..108%, with release build the results could be a bit worse, as assertions are run only once with "fixups/array" scenarios):

<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:///C:/Users/jposwiat/AppData/Local/Temp/msohtmlclip1/01/clip.htm">
<link rel=File-List
href="file:///C:/Users/jposwiat/AppData/Local/Temp/msohtmlclip1/01/clip_filelist.xml">
<style>
<!--table
	{mso-displayed-decimal-separator:"\.";
	mso-displayed-thousand-separator:"\,";}
@page
	{margin:.75in .7in .75in .7in;
	mso-header-margin:.3in;
	mso-footer-margin:.3in;}
tr
	{mso-height-source:auto;}
col
	{mso-width-source:auto;}
br
	{mso-data-placement:same-cell;}
td
	{padding-top:1px;
	padding-right:1px;
	padding-left:1px;
	mso-ignore:padding;
	color:black;
	font-size:11.0pt;
	font-weight:400;
	font-style:normal;
	text-decoration:none;
	font-family:"Aptos Narrow", sans-serif;
	mso-font-charset:0;
	mso-number-format:General;
	text-align:general;
	vertical-align:bottom;
	border:none;
	mso-background-source:auto;
	mso-pattern:auto;
	mso-protection:locked visible;
	white-space:nowrap;
	mso-rotate:0;}
-->
</style>
</head>

<body link="#467886" vlink="#96607D">


single value at once |   |   |   |   |   |   | fixups |   |   |   |   |   | single/fixups
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
loop | default_gen | 401.336 | 400.668 | 399.013 | 400.339 |   |   | 476.938 | 475.356 | 473.516 | 475.27 |   |   | 0.8423401
loop | patterned | 526.894 | 527.902 | 528.448 | 527.748 |   |   | 627.52 | 594.576 | 589.588 | 603.8947 |   |   | 0.8739074
loop | obsolete_gen | 507.398 | 514.283 | 507.123 | 509.60133 |   |   | 572.796 | 565.608 | 562.47 | 566.958 |   |   | 0.8988344
loop | fp_zero | 303.885 | 304.68 | 304.475 | 304.34667 |   |   | 381.258 | 381.049 | 382.047 | 381.4513 |   |   | 0.797865
loop | fp_val_zero | 311.018 | 311.613 | 311.921 | 311.51733 |   |   | 397.122 | 395.189 | 394.728 | 395.6797 |   |   | 0.7872968
loop | normal | 404.236 | 399.653 | 401.326 | 401.73833 |   |   | 506.619 | 505.783 | 505.144 | 505.8487 |   |   | 0.7941868
loop | positive | 308.503 | 307.554 | 308.41 | 308.15567 |   |   | 380.498 | 380.719 | 380.638 | 380.6183 |   |   | 0.8096186
loop | fast_zero | 35.72 | 35.768 | 35.811 | 35.766333 |   |   | 106.784 | 106.419 | 106.166 | 106.4563 |   |   | 0.3359719
loop | fast_memset | 185.974 | 224.735 | 185.54 | 198.74967 |   |   | 208.424 | 207.872 | 208.437 | 208.2443 |   |   | 0.9544061
loop | default_init | 11.169 | 18.955 | 11.169 | 13.764333 |   |   | 11.172 | 11.174 | 11.169 | 11.17167 |   |   | 1.2320752
loop | obsolete_fn | 478.634 | 480.431 | 478.349 | 479.138 |   |   | 478.492 | 478.815 | 478.657 | 478.6547 |   |   | 1.0010098
loop | memset | 11.168 | 11.171 | 11.169 | 11.169333 |   |   | 11.169 | 11.188 | 11.173 | 11.17667 |   |   | 0.9993439
array | default_gen | 388.935 | 390.434 | 389.785 | 389.718 | 0.97347 |   | 359.196 | 361.96 | 359.527 | 360.2277 | 0.7579432 |   | 1.0818658
array | patterned | 518.134 | 519.314 | 518.109 | 518.519 | 0.982512 |   | 519.022 | 520.746 | 519.777 | 519.8483 | 0.8608262 |   | 0.9974428
array | obsolete_gen | 495.98 | 496.604 | 496.706 | 496.43 | 0.974154 |   | 501.894 | 499.832 | 500.383 | 500.703 | 0.8831395 |   | 0.991466
array | fp_zero | 296.979 | 296.843 | 297.351 | 297.05767 | 0.97605 |   | 279.065 | 279.29 | 280.342 | 279.5657 | 0.7329 |   | 1.0625685
array | fp_val_zero | 297.817 | 297.304 | 297.562 | 297.561 | 0.955199 |   | 278.837 | 279.21 | 281.307 | 279.7847 | 0.7070989 |   | 1.0635358
array | normal | 388.676 | 389.018 | 389.831 | 389.175 | 0.968728 |   | 407.499 | 407.286 | 407.22 | 407.335 | 0.8052507 |   | 0.9554175
array | positive | 297.745 | 298.715 | 297.853 | 298.10433 | 0.967382 |   | 277.628 | 276.813 | 276.424 | 276.955 | 0.7276449 |   | 1.0763638
array | fast_zero | 0.707 | 0.731 | 0.716 | 0.718 | 0.020075 |   | 0.71 | 0.708 | 0.711 | 0.709667 | 0.0066663 |   | 1.0117426
array | fast_memset | 66.23 | 66.321 | 66.471 | 66.340667 | 0.33379 |   | 66.747 | 66.625 | 66.283 | 66.55167 | 0.3195845 |   | 0.9968295
</body>
</html>

"Patterned"/"obsolete gen" are only slightly worse than "default_gen" as very slow mantissa generation function is replaced with just single `reandom()`.

thiagomacieira commented on framework/Floats.cpp:
```diff
@@ -304,56 +303,558 @@ float from_hfloat8_emulated(HFloat8 f) {
-{
-    BFloat16 f;
+template<typename I>
+I get_random_value(I range) {
```
It avoids having the type-dependent `false`.

jposwiata committed:
[random_bits] Limited RNG usage functions

Two families of functions to "consume" parts of RNG generated
values: both given number of bits and values in given range.
The interface reduced to handle only get_random_bits31() (for
LCG) and get_random_bits128() (for EAS) (with appropriate
template definitions to handle more types)

Signed-off-by: Jarek Poswiata <jaroslaw.poswiata@intel.com>

jposwiata committed:
Functions to generate appropriate (for the test) random numbers.

Default implementation to fill a range with values (single value
for compatibility in the form of new_random_xxx() family).
Generation is handled with overloaded macro(s) configured with
flags defined in FP_GEN_RANDOM_FLAGS enumeration.
Default implementation keeps slow version of preparing chains
of bits (no functional chages to existing tests, the values are
exactly identical for "old" seed values).
Normalization is available only as a part of generation (with
NORMALIZE flag).
Introduced an option to build a random with half-axis range.

Static vectors with percentage added (and existing functions kept
for compatibility, generators use optimized RNG)

Signed-off-by: Jarek Poswiata <jaroslaw.poswiata@intel.com>

jposwiata committed:
[random_bits] Unit tests to verify random_bits/random_value()

Signed-off-by: Jarek Poswiata <jaroslaw.poswiata@intel.com>

jposwiata committed:
[random_bits] Value generator with more uniform distribution

Appropriate number of random bits is used with value rejection
loop to have better result distribution. Rejection loops runs
up to 7 times, when not succeeded, it degrades to biased modulo
operation.

Signed-off-by: Jarek Poswiata <jaroslaw.poswiata@intel.com>

jposwiata committed:
[random_bits] Fetching bits optimization with TSL

To reduce the number of expensive calls to the RNG, get_random_bits()
implements local state (thread storaged). It significantly reduces
the time spent in (thread local too) Engine.

Signed-off-by: Jarek Poswiata <jaroslaw.poswiata@intel.com>

jposwiata commented:
Nobody is buying it, won't sell.
